File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        smallGrids = [[set() for _ in range(3)] for _ in range(3)]
        row = []
        col = []
        for i in range(9):
            smallGrids.append(set())
            row.append(set())
            col.append(set())
        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    if board[i][j] in smallGrids[i//3][j//3] or board[i][j] in row[i] or board[i][j] in col[j]:
                        return False
                    smallGrids[i//3][j//3].add(board[i][j])
                    row[i].add(board[i][j])
                    col[j].add(board[i][j])
        return True

File: 36-valid-sudoku/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        boolean [][] row = new boolean[9][9];
        boolean [][] col = new boolean[9][9];
        boolean [][] box = new boolean[9][9];
        for(int i = 0;i<9;i++){
            for(int j = 0;j<9;j++){
                if(board[i][j] != '.'){
                    int digit = board[i][j] - '1';
                    int boardIndex = (i / 3) * 3 + (j / 3);
                    if(row[i][digit] || col[digit][j] || box[boardIndex][digit]){
                        return false;
                    }
                    row[i][digit] = true;
                    col[digit][j] = true;
                    box[boardIndex][digit] = true;
                }
            }
        }
        return true;
    }
} 